### PR TITLE
Versioning stage 2. Translate multiple times to support multiple UAP versions.

### PR DESCRIPTION
**Pull request by simeoncran**
This translates but does not yet codegen the multiple versions.

### CODE REVIEW

File: source/LottieToWinComp/Float32.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
```
Comment (simeoncran) [inline source/LottieToWinComp/Float32.cs]: Note to reviewers: this file is not being used yet. I wanted to use it in the translator to replace the "epsilon" values we use to separate keyframes. However it's risky to do that as it changes the codegen output on a lot of the corpus, so I'll turn it on later. For now I'm just capturing the code.

File: source/LottieToWinComp/TranslationIssue.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.Toolkit.Uwp.UI.Lottie.LottieToWinComp
+{
+#if PUBLIC
```
Comment (simeoncran) [inline source/LottieToWinComp/TranslationIssue.cs]: Add a comment to describe the struct.

File: source/LottieToWinComp/UapVersionRange.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.Toolkit.Uwp.UI.Lottie.LottieToWinComp
+{
+    /// <summary>
+    /// Describes a range of UAP versions. A <see cref="Start"/> value of <c>null</c>
+    /// indicates all versions less than or equal to the <see cref="End"/> value.
+    /// An <see cref="End"/> value of <c>null</c> indicates that it affects all versions greater
```
Comment (simeoncran) [inline source/LottieToWinComp/UapVersionRange.cs]: >it affects [](start = 65, length = 10)

Remove "it affects". That was some cruft from an earlier factoring.

File: source/LottieToWinComp/UapVersionRange.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    /// than or equal to the <see cref="Start"/> value. Both values being <c>null</c>
+    /// indicates all version.
```
Comment (simeoncran) [inline source/LottieToWinComp/UapVersionRange.cs]: >version [](start = 22, length = 7)

Typo: versions.

File: source/LottieToWinComp/UapVersionRange.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        /// <summary>
+        /// The start of the range, or <c>null</c> to indicate all values
+        /// less than or equal to <see cref="End"/>.
+        /// </summary>
+        public uint? Start { get; internal set; }
+
+        /// <summary>
+        /// The end of the range, or <c>null</c> to indicate all values
+        /// greater than or equal to <see cref="End"/>.
```
Comment (simeoncran) [inline source/LottieToWinComp/UapVersionRange.cs]: >End [](start = 48, length = 3)

typpo: Start.

File: source/LottieToWinComp/UapVersionRange.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        /// </summary>
+        public uint? End { get; internal set; }
+
+        public override string ToString()
+        {
+            if (Start.HasValue)
+            {
+                if (End.HasValue)
+                {
+                    if (End.Value == Start.Value)
```
Comment (simeoncran) [inline source/LottieToWinComp/UapVersionRange.cs]: >Value [](start = 28, length = 5)

.Value is redundant on these... they both have values.

File: LottieGen/LottieFileProcessor.cs [hunk 7]
```diff
@@ -553,6 +564,46 @@ string IssueToString(string originatingFile, (string Code, string Description) i
...
+    string CreateUapVersionRangeQualifier(UapVersionRange uapVersionRange)
+    {
+        // Ensure ranges are expressed consistently.
+        uapVersionRange.NormalizeForMinimumVersion(_minimumUapVersion);
+
+        if (uapVersionRange.Start.HasValue)
+        {
+            if (uapVersionRange.End.HasValue)
+            {
+                if (uapVersionRange.End.Value == uapVersionRange.Start.Value)
```
Comment (simeoncran) [inline LottieGen/LottieFileProcessor.cs]: >Value [](start = 71, length = 5)

.Value is redundant - they both have values.

File: source/LottieToWinComp/LottieToMultiVersionWinCompTranslator.cs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+using Microsoft.Toolkit.Uwp.UI.Lottie.LottieData;
+
+namespace Microsoft.Toolkit.Uwp.UI.Lottie.LottieToWinComp
+{
+#if PUBLIC
+    public
+#endif
+    sealed class LottieToMultiVersionWinCompTranslator
+    {
+        // We will never produce code that is compatible with code before version 7.
```
Comment (eliezerpMS) [inline source/LottieToWinComp/LottieToMultiVersionWinCompTranslator.cs]: >We will never produce code that is compatible with code before version 7. [](start = 11, length = 73)

Super nit: Reads a bit awkwardly. Maybe something like:
"Lowest version of UAP we will produce code for."

File: source/LottieToWinComp/LottieToMultiVersionWinCompTranslator.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        const uint LowestValidUapVersion = 7;
+
+        /// <summary>
+        /// Attempts to translates the given <see cref="LottieData.LottieComposition"/> for a range of UAP versions,
```
Comment (eliezerpMS) [inline source/LottieToWinComp/LottieToMultiVersionWinCompTranslator.cs]: >translates [](start = 24, length = 10)

change to: Translate

File: source/LottieToWinComp/LottieToMultiVersionWinCompTranslator.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        /// producing one or more translations.
+        /// </summary>
+        /// <param name="lottieComposition">The <see cref="LottieComposition"/> to translate.</param>
+        /// <param name="targetUapVersion">The version of UAP that the translator will ensure compatibility with. Must be >= 7.</param>
+        /// <param name="minimumUapVersion">The lowest version of UAP on which the result must run. Must be >= 7 and &lt;= targetUapVersion.</param>
+        /// <param name="strictTranslation">If true, throw an exception if translation issues are found.</param>
+        /// <param name="addCodegenDescriptions">Add descriptions to objects for comments on generated code.</param>
+        /// <returns>The results of the translation and the issues.</returns>
+        public static (IReadOnlyList<TranslationResult> translationResults, IReadOnlyList<(TranslationIssue, UapVersionRange)> issues) TryTranslateLottieComposition(
```
Comment (eliezerpMS) [inline source/LottieToWinComp/LottieToMultiVersionWinCompTranslator.cs]: >public static (IReadOnlyList<TranslationResult> translationResults, IReadOnlyList<(TranslationIssue, UapVersionRange)> issues) TryTranslateLottieComposition( [](start = 8, length = 157)

Super nit: This line is 158 characters long. I am wondering if there is some formatting guidance we want to settle on for cases like this?

File: source/LottieToWinComp/LottieToWinCompTranslator.cs [hunk 11]
```diff
@@ -2779,10 +2792,26 @@ CompositionColorBrush TranslateSolidColorFill(TranslationContext context, SolidC
...
-            ReportUapVersionCompromiseIssue(8, "CompositionRadialGradientBrush");
+            if (IsUapApiAvailable(nameof(CompositionRadialGradientBrush), versionDependentFeatureDescription: "Radial gradient fill"))
+            {
+                return TranslateRadialGradientFillToCompositionRadialGradientBrush(context, shapeFill, colorStops, opacityPercentValue);
+            }
+            else
+            {
+                // Radial gradient are not available. Convert the RadialGradientFill to a LinearGradientFill as a compromise.
+                var fillArgs = new ShapeLayerContent.ShapeLayerContentArgs() { BlendMode = shapeFill.BlendMode, MatchName = shapeFill.MatchName, Name = shapeFill.Name };
+                var linearFill = new LinearGradientFill(in fillArgs, shapeFill.FillType, shapeFill.OpacityPercent, shapeFill.StartPoint, shapeFill.EndPoint, shapeFill.ColorStops, shapeFill.OpacityPercentStops);
```
Comment (eliezerpMS) [inline source/LottieToWinComp/LottieToWinCompTranslator.cs]: nit: These lines are pretty long. I would not mind if they got shorter.

File: source/LottieToWinComp/LottieToWinCompTranslator.cs [hunk 12]
```diff
@@ -2779,10 +2792,26 @@ CompositionColorBrush TranslateSolidColorFill(TranslationContext context, SolidC
...
-            // TODO - this will never be true once we start translating to version-compensated code. Instead
-            // we will generate something that doesn't require the feature.
-            ReportUapVersionCompromiseIssue(8, "CompositionRadialGradientBrush");
+            if (IsUapApiAvailable(nameof(CompositionRadialGradientBrush), versionDependentFeatureDescription: "Radial gradient fill"))
+            {
+                return TranslateRadialGradientFillToCompositionRadialGradientBrush(context, shapeFill, colorStops, opacityPercentValue);
+            }
+            else
+            {
+                // Radial gradient are not available. Convert the RadialGradientFill to a LinearGradientFill as a compromise.
```
Comment (eliezerpMS) [inline source/LottieToWinComp/LottieToWinCompTranslator.cs]: >Radial gradient are not available [](start = 19, length = 33)

Radial gradient is not available. 
?

File: source/LottieToWinComp/TranslationIssue.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.Toolkit.Uwp.UI.Lottie.LottieToWinComp
+{
+    /// <summary>
+    /// Describes an issue discovered during translation of a Lottie composition.
+    /// The <see cref="Code"/> is an alphanumeric code that identifies the type
+    /// of issue and can be usd to help search online for more information.
```
Comment (eliezerpMS) [inline source/LottieToWinComp/TranslationIssue.cs]: >usd [](start = 28, length = 3)

used

**Pull request merged**